xonsh/proc.py

```diff
@@ -68,51 +68,15 @@ def RE_VT100_ESCAPE():
     return re.compile(b'(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]')
 
 
-def populate_char_queue(reader, fd, queue):
-    """Reads single characters from a file descriptor into a queue.
-    If this ends or fails, it flags the calling reader object as closed.
-    """
-    while True:
-        try:
-            c = os.read(fd, 1)
-        except OSError:
-            reader.closed = True
-            break
-        if c:
-            queue.put(c)
-        else:
-            reader.closed = True
-            break
-
-
-class NonBlockingFDReader:
-    """A class for reading characters from a file descriptor on a background
-    thread. This has the advantages that the calling thread can close the
-    file and that the reading does not block the calling thread.
-    """
+class QueueReader:
```

Comment:
I think it should have a `__init__` function that sets the class methods that the parent class expects to be present. The child class `NonBlockingFDReader` can then overload them if it wants.